Ballerina lang-parser implementation (#15750)
ballerina-platform/ballerina-lang: stdlib/lang-parser/README.md

> ## Ballerina-lang-parser

Reviewer: Shouldn't we have a proper heading here instead?
```suggestion
## Ballerina Parser
```

And then have a small description to say this is a parser written for Ballerina in Ballerina?

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

> The parser is implemented following the LL(K) recursive Descent parsing pattern. This pattern analyses the syntactic structure using k>1 lookahead tokens.

Reviewer: ```suggestion
The parser is implemented following the LL(K) recursive descent parsing pattern. This pattern analyses the syntactic structure using k>1 lookahead tokens.
```

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

**Token Structure**
```ballerina
Token record {
    int tokenType;
    string text;
    int startPos;
    int endPos;
    int lineNumber;
    int index;
    string? whiteSpace;
}
```

> Grammar supported by the current implementation is given below.

Reviewer: Let's move this to after the heading `**Parser Grammar**` and say 
```suggestion
The current implementation supports the following grammar.
```

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

**Parser Grammar**
```antlrv4
Definition
    | Function definition
Function definition
    | FUNCTION <callable Unit Signature> <callable Unit body>
Callable Unit Signature
    | IDENTIFIER ()
Callable Unit Body
    | { <statement*>}
Statement
    |<variable definition statement>
    |<continue statement>
Variable definition statement
    | <valueTypeName> IDENTIFIER SEMICOLON
    | <valueTypeName>  IDENTIFIER ASSIGN <expression> SEMICOLON
Continue statement
    | CONTINUE SEMICOLON
valueTypeName
    | INT
Expression
    | <simple literal>
    | <variable reference>
    | expression (DIVISION | MULTIPLICATION | MOD) expression
    | expression ( ADD | SUB ) expression
    | expression (LT_EQUAL | GT_EQUAL | GT | LT)expression
    | expression (EQUAL | NOT_EQUAL) expression
    | expression (REF_EQUAL | REF_NOT_EQUAL) expression
    | (ADD | SUB | NOT | BIT_COMPLEMENT | UNTAINT) expression
    | <tuple literal>
    
Simple Literal
    | IntegerLiteral //[0-9]
    | EmptyTupleLiteral

Variable reference
    | IDENTIFIER

EmptyTupleLiteral
    | LBRACE RBRACE

TupleLiteral
     |LEFT_PARENTHESIS expression (COMMA expression)* RIGHT_PARENTHESIS
```

#### Testing the parser

> 1. Navigate to the <PROJECT_ROOT_DIRECTORY>/inputFile.txt and add the source code that is needed to be tested.

Reviewer: What if we change the `main` function to have a string parameter `fileName`, and accept the name of the file instead of asking to add in a predefined file (which IMO should have the `.bal` extension instead of `.txt`). 

That way we can ask the user to just introduce the file in the `PROJECT_ROOT_DIRECTORY`

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

	Sample source code:
	```ballerina

> 	function foo(){

Reviewer: Shall we format this properly?

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

    	int a = 2 + (3*5);
    }
	```

> 2. Run the `ballerina run parserModule` command from the <PROJECT_ROOT_DIRECTORY>

Reviewer: Instead of `parserModule`, shall we just use `parser`?

If we also do the change to accept the file name as an argument we can change this to

```suggestion
2. Run the following command from the <PROJECT_ROOT_DIRECTORY>
`$ ballerina run parser test_file.bal`
```

> The built AST will be printed as a Record.

Reviewer: record or JSON?

Author: Ballerina lang-parser implementation in no longer maintained and has already being removed from the ballerina-lang repo.

							"lineNumber": 2,
							"index": 15,
							"whiteSpace": null
						}, {
							"tokenType": 6,
							"text": "(",
							"startPos": 17,
							"endPos": 17,
							"lineNumber": 2,
							"index": 11,
							"whiteSpace": " "
						}],
						"tupleExprList": [{
							"nodeKind": "binaryExpression",
							"tokenList": [{
								"tokenType": 15,
								"text": "*",
								"startPos": 19,
								"endPos": 19,
								"lineNumber": 2,
								"index": 13,
								"whiteSpace": null
							}],
							"operatorKind": "*",
							"leftExpr": {
								"nodeKind": "integerLiteral",
								"tokenList": [{
									"tokenType": 47,
									"text": "3",
									"startPos": 18,
									"endPos": 18,
									"lineNumber": 2,
									"index": 12,
									"whiteSpace": null
								}],
								"number": "3"
							},
							"rightExpr": {
								"nodeKind": "integerLiteral",
								"tokenList": [{
									"tokenType": 47,
									"text": "5",
									"startPos": 20,
									"endPos": 20,
									"lineNumber": 2,
									"index": 14,
									"whiteSpace": null
								}],
								"number": "5"
							}
						}]
					}
				}
			}]
		}
	}]
}

```

#### Unit Tests

> To execute the Unit tests, run the following command from the <PROJECT_ROOT_DIRECTORY>

Reviewer: ```suggestion
To execute the unit tests, run the following command from the <PROJECT_ROOT_DIRECTORY>
```

 `ballerina test parserModule` 
 

>  serialized JSON outputs for the unit tests can be found below.

Reviewer: ```suggestion
 Serialized JSON output for each test is available at
```

> 1. [function rule](resources/function/output(AST)).

Reviewer: Do we need the fullstops at the end?
